embulk-executor-mapreduce/src/main/java/org/embulk/executor/mapreduce/MapReduceExecutor.java

```diff
@@ -393,23 +396,42 @@ private void updateProcessState(Job job, TaskReportSet reportSet, Path stateDir,
             }
         }
 
+        List<RemoteTaskFailedException> aggregated = new LinkedList<>();
+
         for (AttemptReport report : reportSet.getLatestInputAttemptReports()) {
-            updateTaskState(state.getInputTaskState(report.getInputTaskIndex().get()), report.getAttemptState(), true);
+            updateTaskState(state.getInputTaskState(report.getInputTaskIndex().get()),
+                    report.getAttemptState(), true, aggregated, inProgress);
         }
 
         for (AttemptReport report : reportSet.getLatestOutputAttemptReports()) {
-            updateTaskState(state.getOutputTaskState(report.getOutputTaskIndex().get()), report.getAttemptState(), false);
+            updateTaskState(state.getOutputTaskState(report.getOutputTaskIndex().get()),
+                    report.getAttemptState(), false, aggregated, inProgress);
+        }
+
+        if (!inProgress && !aggregated.isEmpty()) {
+            DataException de = new DataException("aggregate DataExceptions on mappers and reducers");
+            for (RemoteTaskFailedException e : aggregated) {
+                de.addSuppressed(e);
+            }
+            throw de;
         }
     }
 
-    private static void updateTaskState(TaskState state, AttemptState attempt, boolean isInput)
+    private static void updateTaskState(TaskState state, AttemptState attempt, boolean isInput,
+            List<RemoteTaskFailedException> aggregated, boolean inProgress)
     {
         state.start();
         Optional<TaskReport> taskReport = isInput ? attempt.getInputTaskReport() : attempt.getOutputTaskReport();
         boolean committed = taskReport.isPresent();
         if (attempt.getException().isPresent()) {
             if (!state.isCommitted()) {
-                state.setException(new RemoteTaskFailedException(attempt.getException().get()));
+                RemoteTaskFailedException e = new RemoteTaskFailedException(attempt.getException().get());
+                state.setException(e);
+
+                if (!inProgress && e.getMessage().startsWith(DataException.class.getName())) {
```

Comment:
Using matching to a string looks bad idea. It's less reliable (less tolerant from future changes).